feat: funnel chart

magnew opened the pull request:
<!-- Thanks so much for your PR, your contribution is appreciated! ❤️ -->

Closes: #2594 

### Description:

Adds a funnel chart viz (from eCharts). This is a draft for now but it's not far from being ready. At the moment you can pick which column in the data to use for the steps. Customers requested also to be able to use a row. We can think about other options (colors?). 

Here's what it looks like for now.

![Kapture 2024-06-17 at 19 51 26](https://github.com/lightdash/lightdash/assets/1864179/62dccb15-5775-4eb8-81b5-5d5072a5698a)


### Reviewer actions

- [ ] I have manually tested the changes in the preview environment
- [ ] I have reviewed the code
- [ ] I understand that "request changes" will block this PR from merging

hamzahc1 commented:
Colours would definitely help here - a few other suggestions: 
- Hide or show label?
- Change label position? (inside - left - top)

magnew commented:
> Colours would definitely help here - a few other suggestions:
> 
> * Hide or show label?
> * Change label position? (inside - left - top)

@hamzahc1 

Totally doable. What do you think the priority for colors is? Being able to specify a color for each step? Or different auto coloring?

I can add those label options.

hamzahc1 commented:
Likely pulling from the colour scheme initially if it’s set. Otherwise
changing each step manually has the most control. At some point better auto
colouring but feels like a nice to have.

On Wed, 19 Jun 2024 at 09:43, magnew ***@***.***> wrote:

> Colours would definitely help here - a few other suggestions:
>
>    - Hide or show label?
>    - Change label position? (inside - left - top)
>
> @hamzahc1 <https://github.com/hamzahc1>
>
> Totally doable. What do you think the priority for colors is? Being able
> to specify a color for each step? Or different auto coloring?
>
> I can add those label options.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/lightdash/lightdash/pull/10445#issuecomment-2178108442>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADNNJJIIFTUMTITK62QWGI3ZIFAEDAVCNFSM6AAAAABJOO6HZ2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCNZYGEYDQNBUGI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

magnew commented:
@IrakliJani and/or @joaoviana, this is pretty much ready for review. I've fixed the blockers Katie had and I will follow up with the config options Hamzah is requesting. 

I added this dashboard to the render deployment to show the two different data formats: https://lightdash-pr-10445.onrender.com/projects/3675b69e-8324-4110-bdca-059031aa8da3/dashboards/c8333ddc-206f-4b9a-8960-06d2794922d8/view

Let me know if you have any questions!

magnew commented on packages/frontend/src/components/FunnelChart/FunnelChartContextMenu.tsx:
This is almost entirely the same as PieChartContextMenu. I want to unify them, but I want to do that after this review so I'm not editting the pie chart here.

magnew replied:
I won't argue 😄. It was a place I was thinking about reducing repetition, but it's true that they are diverging mroe.

hamzahc1 commented:
Not relevant for this PR but when using columns as the steps - not being able to easily re-order the columns in the table or in the chart config could be quite annoying

joaoviana commented:
Early feedback (you can record these for future tickets, non-blocking)

- When there's no metric selected I would like to see an empty state
<img width="1894" alt="image" src="https://github.com/lightdash/lightdash/assets/7611706/81e67e3f-d789-432d-abc1-6ac89358fdcc">

- When I have data orientation as `columns` and I hover on the chart and click `copy value`, the value copied isn't the one I'm hovering on, but its actual `value` - I guess that makes sense? 
<img width="1918" alt="image" src="https://github.com/lightdash/lightdash/assets/7611706/21440d65-09ac-48a9-bfd5-2d3ff88fe88c">

- Sorts don't affect the funnel chart viz(?) - I think I would expect it to flip if I sort `Order status` in descending alphabetical order? 🤷

joaoviana commented:
**WIP, will edit this as I go**

(non-blocking)
`getCreateEventProperties` in `SavedChartService` should include `funnel`


Tested user scenarios
Saving & Dashboard interaction
1. Added a saved funnel chart to dashboard and the filter in dashboards was applied correctly
2. Then went to `explore from here` and that also worked
3. Turned off the filter and saw the visualization change
Saving `explore from here` and directly into a dashboard
- all worked well
Image exports
- Exports in saved explorer
- Dashboard exports
- Scheduled deliveries
View underlying data 
- works as expected for both `data orientation` types

magnew commented:
> Not relevant for this PR but when using columns as the steps - not being able to easily re-order the columns in the table or in the chart config could be quite annoying

@hamzahc1 Agreed. @joaoviana, you also mentioned sorting. We can re-visit this, but at the moment it is set up to always sort in descending order. This is actually the default for the echarts funnel and I think it's how many of them work. We could add sorting options, but this is fairly easy to explain/justify.

joaoviana reviewed (commented):
Great work @magnew ! 
LMK what you think of the comments I've left

Also

(non-blocking)
`getCreateEventProperties` in `SavedChartService` should include `funnel`

IrakliJani commented on packages/frontend/src/components/FunnelChart/index.tsx:
```diff
@@ -0,0 +1,135 @@
+        <SuboptimalState
+            title="Loading chart"
+            loading
+            className="loading_chart"
```
I believe this is for scheduled deliveries?

magnew replied:
The classname? Does that mean it should come out?

joaoviana replied:
Correct, this is for scheduled deliveries. We should keep it so we can keep track of the loading spinners

IrakliJani commented on packages/frontend/src/hooks/useFunnelChartConfig.ts:
```diff
@@ -0,0 +1,178 @@
+    );
+
+    // The value at the top of the funnel.
+    const maxValue = useRef(0);
```
why do we have this?

magnew replied:
This is the value that the funnel 'starts' with. Percentages in later stages are based on this number. I was surprised how bad a job echarts does here -- we have to calculate percentages ourselves to get it right. 

I will make the comment more clear

IrakliJani reviewed (commented):
code looks solid! I added a few questions but did not have the chance to test the UI.

magnew commented on packages/frontend/src/components/Explorer/VisualizationCardOptions/index.tsx:
```diff
@@ -163,6 +165,11 @@ const VisualizationCardOptions: FC = memo(() => {
+            case ChartType.FUNNEL:
+                return {
+                    text: 'Funnel chart',
+                    icon: <MantineIcon icon={IconFilter} color="gray" />,
```
Haha, I was looking for that one and didn't see it. Thanks, will update.

magnew replied:
Ok, part of why I didn't find that is it's not in our tabler version. I'm going to leave it as the funnel for now and file a ticket to update tabler.

magnew replied:
https://github.com/lightdash/lightdash/issues/10475
https://github.com/lightdash/lightdash/issues/10476

magnew commented on packages/frontend/src/components/FunnelChart/index.tsx:
```diff
@@ -0,0 +1,135 @@
+} from './FunnelChartContextMenu';
+
+const EmptyChart = () => (
+    <div style={{ height: '100%', width: '100%', padding: '50px 0' }}>
```
Yup, sure.

magnew commented:
> Great work @magnew ! LMK what you think of the comments I've left
> 
> Also
> 
> (non-blocking) `getCreateEventProperties` in `SavedChartService` should include `funnel`

Good call. I missed that.

magnew commented on packages/frontend/src/hooks/useFunnelChartConfig.ts:
```diff
@@ -0,0 +1,178 @@
+                };
+            });
+        } else {
+            return allNumericFieldIds.reduce((acc, id) => {
```
Done, although I see these as basically the exact same amount of code :D

magnew commented:
> Early feedback (you can record these for future tickets, non-blocking)
> 
> * When there's no metric selected I would like to see an empty state
> 
> <img alt="image" width="1894" src="https://private-user-images.githubusercontent.com/7611706/341378862-81e67e3f-d789-432d-abc1-6ac89358fdcc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg4OTgzMjQsIm5iZiI6MTcxODg5ODAyNCwicGF0aCI6Ii83NjExNzA2LzM0MTM3ODg2Mi04MWU2N2UzZi1kNzg5LTQzMmQtYWJjMS02YWM4OTM1OGZkY2MucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYyMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MjBUMTU0MDI0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MGQ1NjZkNWIyZmY2NGJjNTAzZjc3ZTkzNGI2MzllOWQ3YmYzNDZiMDUwOGY3MTAwMTU3NGVjZjJkMjc2MzQzMCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.5tu2uH5pAwy32txTESkYgpE3xpC5mKgyMlNu4s3EjYo">
> * When I have data orientation as `columns` and I hover on the chart and click `copy value`, the value copied isn't the one I'm hovering on, but its actual `value` - I guess that makes sense?
> 
> <img alt="image" width="1918" src="https://private-user-images.githubusercontent.com/7611706/341379410-21440d65-09ac-48a9-bfd5-2d3ff88fe88c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg4OTgzMjQsIm5iZiI6MTcxODg5ODAyNCwicGF0aCI6Ii83NjExNzA2LzM0MTM3OTQxMC0yMTQ0MGQ2NS0wOWFjLTQ4YTktYmZkNS0yZDNmZjg4ZmU4OGMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYyMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MjBUMTU0MDI0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZDRlYjYzZjJkNWM2YTkxYjFlNGNjYzcwMWEwODU5MzA3MDBlYTAxMGNmOTU1MDE1NmVlYzUyMDNlMDljNTJjMCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.UInp-dvuadKfv4tKbqjYJrgbK7YCQ6dHr2X8ZjRW1PY">
> * Sorts don't affect the funnel chart viz(?) - I think I would expect it to flip if I sort `Order status` in descending alphabetical order? 🤷

Fixed the empty state. I think the copy issue is correct? You get the value instead of the label. We already talked about sorting.

magnew commented on packages/frontend/src/components/LightdashVisualization/VisualizationConfigFunnel.tsx:
```diff
@@ -0,0 +1,88 @@
+        const tableCalculations = getTableCalculationsFromItemsMap(itemsMap);
+        return {
+            dimensions: getDimensionsFromItemsMap(itemsMap ?? {}),
+            numericFields: { ...metrics, ...tableCalculations },
```
Added a numeric fitler here. I'll look into pie charts.

magnew commented:
@joaoviana and @IrakliJani I think I have addressed all of the feedback so far. Thank you for the good reviews!

joaoviana reviewed (approved):
Tested Table calculations
- Created a number
- added to the funnel chart
- the changed its type to `string` 
- and it de-selected that tc and found a metric and selected it - great!

Approving as the improvements and other considerations will be done in different PRs later
